File: leetcode-problems/0001-two-sum/solution.ts
function twoSum(nums: number[], target: number): number[] {
    // Key: the number itself, Value: its index in the array
    const seenNumbers = new Map<number, number>();

    for (let i = 0; i < nums.length; i++) {
        const currentNum = nums[i];
        const complement = target - currentNum;

        // Check if the number we need has already been seen
        if (seenNumbers.has(complement)) {
            // Return the index of the complement and the current index
            return [seenNumbers.get(complement), i];
        }

        // Otherwise, store the current number and its index for future reference
        seenNumbers.set(currentNum, i);
    }

    // Return an empty array if no solution is found (per standard safety)
    return [];
}

File: LeetCode-Problems/0001-two-sum/solution.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen_numbers = {}

        # enumerate() gives us both the index and the value instantly
        for i, current_num in enumerate(nums):
            complement = target - current_num

            # Dictionary lookups in Python are O(1) on average
            if complement in seen_numbers:
                return [seen_numbers[complement], i]

            seen_numbers[current_num] = i

        return []
